### DISCUSSION TITLE
Bash shell options for errors

### DISCUSSION THREAD

**Discussion by adthrasher** [stjude-rust-labs/wdl#60] [Rule Proposals]
We often use some variation of `set -euo pipefail` in our task declarations. It would be helpful to have a lint rule to catch if this is forgotten.

For completeness here is the BASH [man page](https://www.gnu.org/software/bash/manual/html_node/The-Set-Builtin.html
) for `set`.

> -e
Exit immediately if a pipeline (see [Pipelines](https://www.gnu.org/software/bash/manual/html_node/Pipelines.html)), which may consist of a single simple command (see [Simple Commands](https://www.gnu.org/software/bash/manual/html_node/Simple-Commands.html)), a list (see [Lists of Commands](https://www.gnu.org/software/bash/manual/html_node/Lists.html)), or a compound command (see [Compound Commands](https://www.gnu.org/software/bash/manual/html_node/Compound-Commands.html)) returns a non-zero status. The shell does not exit if the command that fails is part of the command list immediately following a while or until keyword, part of the test in an if statement, part of any command executed in a && or || list except the command following the final && or ||, any command in a pipeline but the last, or if the command’s return status is being inverted with !. If a compound command other than a subshell returns a non-zero status because a command failed while -e was being ignored, the shell does not exit. A trap on ERR, if set, is executed before the shell exits.
>
> This option applies to the shell environment and each subshell environment separately (see [Command Execution Environment](https://www.gnu.org/software/bash/manual/html_node/Command-Execution-Environment.html)), and may cause subshells to exit before executing all the commands in the subshell.
>
> If a compound command or shell function executes in a context where -e is being ignored, none of the commands executed within the compound command or function body will be affected by the -e setting, even if -e is set and a command returns a failure status. If a compound command or shell function sets -e while executing in a context where -e is ignored, that setting will not have any effect until the compound command or the command containing the function call completes.
>
>-o option-name
>Set the option corresponding to option-name:
>pipefail
If set, the return value of a pipeline is the value of the last (rightmost) command to exit with a non-zero status, or zero if all commands in the pipeline exit successfully. This option is disabled by default.
>
>-u
Treat unset variables and parameters other than the special parameters ‘@’ or ‘*’, or array variables subscripted with ‘@’ or ‘*’, as an error when performing parameter expansion. An error message will be written to the standard error, and a non-interactive shell will exit.

**Comment by a-frantz** [comment]
Many WDL tasks are single (pipeless) commands. We should _not_ warn about a missing `set -euo pipefail` in those cases. That means we can't just blindly check that the string is present in the beginning of a command block, we will have to do some actual Bash parsing. That strikes me as something we want to avoid. I don't want to parse Bash.

I think this can tie into a future [shellcheck](https://www.shellcheck.net/#) call. I have not gone through the shellcheck docs, but they may have a warning that accomplishes the above more robustly than any implementation from us. If not, I'm sure someone has wanted this lint before and we can find a solution off the shelf somewhere.

Basically, I think this is a lint we outsource as opposed to writing our own. But I have not searched yet to see whether there's tooling available for us.
